TransformFusion/Logic/vtkSlicerTransformFusionLogic.cxx

```diff
@@ -38,133 +38,381 @@
+void vtkSlicerTransformFusionLogic::PrintSelf( ostream& os, vtkIndent indent )
 {
-  this->Superclass::PrintSelf(os, indent);
+  this->Superclass::PrintSelf( os, indent );
 }
 
 //-----------------------------------------------------------------------------
 void vtkSlicerTransformFusionLogic::RegisterNodes()
 {
   vtkMRMLScene* scene = this->GetMRMLScene();
-  if (scene == NULL)
+  if ( scene == NULL )
   {
-    vtkErrorMacro("Null scene");
+    vtkErrorMacro( "Null scene" );
     return;
   }
   
-  scene->RegisterNodeClass(vtkSmartPointer<vtkMRMLTransformFusionNode>::New());
+  scene->RegisterNodeClass( vtkSmartPointer<vtkMRMLTransformFusionNode>::New() );
 }
 
-//----------------------------------------------------------------------------
-void vtkSlicerTransformFusionLogic::SetAndObserveTransformFusionNode(vtkMRMLTransformFusionNode *node)
+//---------------------------------------------------------------------------
+void vtkSlicerTransformFusionLogic::OnMRMLSceneNodeAdded( vtkMRMLNode* node )
+{
+  if ( node == NULL || this->GetMRMLScene() == NULL )
+  {
+    vtkWarningMacro( "OnMRMLSceneNodeAdded: Invalid MRML scene or node" );
+    return;
+  }
+
+  vtkMRMLTransformFusionNode* pNode = vtkMRMLTransformFusionNode::SafeDownCast( node );
+  if ( pNode )
+  {
+    vtkDebugMacro( "OnMRMLSceneNodeAdded: Module node added." );
+    vtkUnObserveMRMLNodeMacro( pNode ); // Remove previous observers.
+    vtkNew<vtkIntArray> events;
+    events->InsertNextValue( vtkCommand::ModifiedEvent );
+    events->InsertNextValue( vtkMRMLTransformFusionNode::InputDataModifiedEvent );
+    vtkObserveMRMLNodeEventsMacro( pNode, events.GetPointer() );
+  }
+}
+
+//---------------------------------------------------------------------------
+void vtkSlicerTransformFusionLogic::OnMRMLSceneNodeRemoved( vtkMRMLNode* node )
 {
-  vtkSetAndObserveMRMLNodeMacro(this->TransformFusionNode, node);
+  if ( node == NULL || this->GetMRMLScene() == NULL )
+  {
+    vtkWarningMacro( "OnMRMLSceneNodeRemoved: Invalid MRML scene or node" );
+    return;
+  }
+
+  if ( node->IsA( "vtkMRMLTransformFusionNode" ) )
+  {
+    vtkDebugMacro( "OnMRMLSceneNodeRemoved" );
+    vtkUnObserveMRMLNodeMacro( node );
+  }
 }
 
 //-----------------------------------------------------------------------------
-void vtkSlicerTransformFusionLogic::fuseInputTransforms(int fusionTechnique) //enum techniqueTypes
+void vtkSlicerTransformFusionLogic::ProcessMRMLNodesEvents( vtkObject* caller, unsigned long event, void* callData )
 {
-  switch (fusionTechnique){
-    case MODE_QUATERNION_AVERAGE:
+  vtkMRMLTransformFusionNode* paramNode = vtkMRMLTransformFusionNode::SafeDownCast( caller );
+  if ( paramNode == NULL )
+  {
+    return;
+  }
+
+  if ( event==vtkMRMLTransformFusionNode::InputDataModifiedEvent )
+  {
+    if ( paramNode->GetUpdateMode() == vtkMRMLTransformFusionNode::UPDATE_MODE_AUTO )
     {
-      this->QuaternionAverageFusion();
-      break;
+      this->Fuse( paramNode );
```

Comment:
If the method updates the output transform then UpdateOutputTransform is a better name. If you compute fuse transforms and return it from the method then GetFusedTransform would be a good name. FuseTransforms is not specific enough, it's not clear what happens when you call the method.